src/query/api/v1/handler/prometheus/remote/attribution.go

```diff
@@ -20,45 +22,81 @@ import (
  */
 
 var errNoOption = errors.New("no option configured for promAttributionMetrics")
-var errInvalidFilter = errors.New("invalid filter configured for promAttributionMetrics, must be `label=value`")
+var errInvalidMatcher = errors.New("invalid matcher configured for promAttributionMetrics")
+
+type matchOp uint8
+
+const (
+	Eq matchOp = iota // ==
+	Ne                // !=
+)
+
+type matcher struct {
+	op      matchOp
+	pattern string
+}
+
+func newMatcher(config string) (string, *matcher, error) {
+	eqRegexp := regexp.MustCompile(`(\w+)==(.*)`)
+	if eqRegexp.MatchString(config) {
+		groups := eqRegexp.FindStringSubmatch(config)
+		return groups[1], &matcher{op: Eq, pattern: groups[2]}, nil
+	}
+	neRegexp := regexp.MustCompile(`(\w+)!=(.*)`)
+	if neRegexp.MatchString(config) {
+		groups := neRegexp.FindStringSubmatch(config)
+		return groups[1], &matcher{op: Ne, pattern: groups[2]}, nil
+	}
+	return "", nil, errInvalidMatcher
+}
+
+func (m *matcher) match(value string) bool {
+	switch m.op {
+	case Eq:
+		return m.pattern == value
+	case Ne:
+		return m.pattern != value
+	default:
+		return false
+	}
+}
 
 type promAttributionMetrics struct {
 	opts               *instrument.AttributionConfiguration
-	filters            map[string]string
+	matchers           map[string]*matcher
 	baseScope          tally.Scope
 	attributedCounters sync.Map
 	counterSize        int32
 	skipSamples        tally.Counter
 	missSamples        tally.Counter
 }
 
-func (pam *promAttributionMetrics) filter(label prompb.Label) bool {
-	if pattern, ok := pam.filters[string(label.Name)]; ok {
-		if pattern == string(label.Value) {
-			return true
+func (pam *promAttributionMetrics) match(labels []prompb.Label) bool {
+	for _, l := range labels {
+		name := string(l.Name)
+		if m, ok := pam.matchers[name]; ok {
+			if !m.match(string(l.Value)) {
+				// if 1 of the matches violates, skip the attribution
+				return false
+			}
 		}
-		// TODO: support regex match
 	}
-	return false
+	return true
 }
 
 func (pam *promAttributionMetrics) attribute(ts prompb.TimeSeries) {
-	if rand.Float64() >= pam.opts.SamplingRate {
+	if rand.Float64() >= pam.opts.SamplingRate || !pam.match(ts.Labels) {
 		return
 	}
-	labelValues := make([]string, len(pam.opts.Labels))
+	matchedValues := make([]string, len(pam.opts.Labels))
 	found := 0
 	sample_count := int64(len(ts.Samples))
 	for _, l := range ts.Labels {
-		if pam.filter(l) {
-			// filter out samples not qualified for this attribution
-			return
-		}
 		labelName := string(l.Name)
 		for i, label := range pam.opts.Labels {
 			if labelName == label {
```

Comment:
This a linear search. Can we construct a map for faster look-ups? If a ts label can match multiple opts.label, the value of the map has to be a slice.

Reply:
Correct, the reason i didn't try to optimize is for attribution, we shouldn't allow a lot of labels, currently we are only using 1 like, another way is to restrict the length when load the configuration:
```
        [if enableAggregator then "attributions"]: [
          {
            name: "cp",
            capacity: 30,
            samplingRate: 1.0,
            labels: ["job"],
            matchers: ["proxy!=kinesis2prom"],
          },
          {
            name: "dp",
            capacity: 100,
            samplingRate: 1.0,
            labels: ["project"],
            matchers: ["proxy==kinesis2prom"],
          },
          {
            name: "pods",
            capacity: 100,
            samplingRate: 1.0,
            labels: ["kubernetes_namespace"],
            matchers: ["job==kubernetes-pods"]
          },
          {
            name: "ksm",
            capacity: 50,
            samplingRate: 1.0,
            labels: ["namespace"],
            matchers: ["job==kube-state-metrics"]
          },
        ],
```